modules/cryptofuncs/cryptofuncs.c

```diff
@@ -94,7 +94,12 @@ tf_hash_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent, gint
       return FALSE;
     }
   state->length = length;
-  md = EVP_get_digestbyname(strcmp(argv[0], "hash") == 0 ? "sha256" : argv[0]);
+  const char *digest_name = strcmp(argv[0], "hash") == 0 ? "sha256" : argv[0];
+  if (strcmp(digest_name, "md4") == 0)
+    {
+      msg_warning_once("WARNING: MD4 hash template function is deprecated and will be removed in future versions");
```

Comment:
Sorry, I didn't review the whole branch, I just have this one comment:

I think we should send a deprecation warning if syslog-ng compiled with Openssl 3 too (and nost just in case with Openssl 1.1):
We could add a dummy template function (.e.g a TEMPLATE_FUNCTION_SIMPLE for md4 in case of Openssl 3.
There is also an example how to do error reporting in case of deprecation:

https://github.com/syslog-ng/syslog-ng/blob/f69ec0a98b34f43820df3157a95d8f52b6b50c63/modules/cryptofuncs/cryptofuncs.c#L97-L102

Thanks @MrAnno to confirm my idea.

Reply:
I've checked the updated version, looks good to me, thanks!